File: site/docs/tbdex/tests/wallet/placeOrder.test.js
import { test, expect, describe, beforeAll, afterAll } from 'vitest';
import { TbdexHttpClient, DevTools, Quote, Order, OrderStatus, Close, Message } from '@tbdex/http-client';
import { DidDht } from '@web5/dids';
import { setupServer } from 'msw/node'
import { http, HttpResponse } from 'msw'

let pfiDid;
let customerDid;
let server;
let order;
let orderStatusMsg = 'Processing';
let closeReason = 'Transaction complete';

describe('Wallet: Place Order', () => {

  beforeAll(async () => {
    customerDid = await DidDht.create({
      options: { publish: true }
    })

    pfiDid = await DidDht.create({
      options:{
        services : [{
          type            : 'PFI',
          id              : 'pfi',
          serviceEndpoint : 'https://localhost:9000'
        }]
      }
    })

    order = Order.create({
      metadata: {
        from       : customerDid.uri,
        to         : pfiDid.uri,
        exchangeId : Message.generateId('rfq'),
        protocol   : "1.0"
      }
    })
    await order.sign(customerDid);

    const orderStatus = OrderStatus.create({
      metadata: {
        from: pfiDid.uri,
        to: customerDid.uri,
        exchangeId: order.exchangeId,
        protocol   : "1.0"
      },
      data: {
        orderStatus: orderStatusMsg
      }
    });
    await orderStatus.sign(pfiDid);

    const close = Close.create({
      metadata: {
        from: pfiDid.uri,
        to: customerDid.uri,
        exchangeId: order.exchangeId,
        protocol   : "1.0"
      },
      data: { 
        reason: closeReason,
        success: true
      }
    });
    await close.sign(pfiDid);

    // Mock responses from the PFI
    server = setupServer(
      http.put(new RegExp('https://localhost:9000/exchanges/(.+)'), () => {
        return HttpResponse.json({ status: 202 })
      }),
      http.post(new RegExp('https://localhost:9000/exchanges/(.+)'), () => {
        return HttpResponse.json({ status: 202 })
      }),
      http.get(new RegExp('https://localhost:9000/exchanges/(.+)'), () => {
        return HttpResponse.json(
          {data: [orderStatus, close]},
          { status: 202 },
        )
      }),
    )
    server.listen({onUnhandledRequest: 'bypass'})
  });

  afterAll(() => {
    server.resetHandlers()
    server.close()
  });

  test('send Order message', async () => {
      const quote = Quote.create({
        metadata: {
          exchangeId : Message.generateId('rfq'),
          from: pfiDid.uri,
          to: customerDid.uri,
          protocol: "1.0"
        },
        data: DevTools.createQuoteData()
      })

      // :snippet-start: createOrderJS
      const order = Order.create({
        metadata: {
          from: customerDid.uri,         // Customer's DID
          to: quote.metadata.from,       // PFI's DID
          exchangeId: quote.exchangeId,  // Exchange ID from the Quote
          protocol: "1.0"                // Version of tbDEX protocol you're using
        }
      });
      // :snippet-end:

      // :snippet-start: signOrderJS
      await order.sign(customerDid);
      // :snippet-end:

      try{
      // :snippet-start: sendOrderJS
      await TbdexHttpClient.submitOrder(order);
      // :snippet-end:
      }catch(e){
        expect.fail(e.message);
      }
  });

  test('listen for Order Status updates', async () => {
    // :snippet-start: listenForOrderStatusJS
    let orderStatusUpdate;
    let close;

    while (!close) {
      const exchange = await TbdexHttpClient.getExchange({
        pfiDid: order.metadata.to,
        did: customerDid,
        exchangeId: order.exchangeId
      });

      for (const message of exchange) {
        if (message instanceof OrderStatus) {
          // a status update to display to your customer
          orderStatusUpdate = message.data.orderStatus;
        }
        else if (message instanceof Close){
          // final message of exchange has been written
          close = message;
          break;
        }
      }
    }
    // :snippet-end:
    expect.soft(orderStatusUpdate).toBe(orderStatusMsg);

    // :snippet-start: getCloseReasonJS
    const isSuccessful = close.data.success;
    const reasonForClose = close.data.reason;
    // :snippet-end:
    expect(reasonForClose).toBe(closeReason);
    expect(isSuccessful).toBe(true);
  });
});

File: site-new/testsuites/testsuite-kotlin/src/test/kotlin/docs/tbdex/wallet/PlaceOrderTest.kt
package website.tbd.developer.site.docs.tbdex.wallet

import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import tbdex.sdk.httpclient.TbdexHttpClient
import tbdex.sdk.httpclient.models.TbdexResponseException
import tbdex.sdk.protocol.models.Close
import tbdex.sdk.protocol.models.Order
import tbdex.sdk.protocol.models.OrderStatus
import tbdex.sdk.protocol.serialization.Json
import java.net.HttpURLConnection
import website.tbd.developer.site.docs.utils.*


/**
 * Tests for Wallet: Placing Orders guide
 */
class PlaceOrderTest {

    private val pfi = TestData.PFI_DID
    private val customerDid = TestData.ALICE_DID
    private lateinit var server: MockWebServer
    private lateinit var getExchangeResponse: String

    @BeforeEach
    fun setup(){
        server = MockWebServer()
        server.start(9000)
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `send Order message`() {
        server.enqueue(MockResponse().setResponseCode(HttpURLConnection.HTTP_OK))

        val quote = TestData.getQuote()
        // :snippet-start: createOrderKt
        val order = Order.create(
            from = customerDid.uri,                 // Customer's DID
            to = quote.metadata.from,               // PFI's DID
            exchangeId = quote.metadata.exchangeId, // Exchange ID from the Quote
            protocol= "1.0"                        // Version of tbDEX protocol you're using
        )
        // :snippet-end:

        // :snippet-start: signOrderKt
        order.sign(customerDid)
        // :snippet-end:

        try {
            // :snippet-start: sendOrderKt
            TbdexHttpClient.submitOrder(order)
            // :snippet-end:
        } catch (e: TbdexResponseException) {
            Assertions.fail("Failed to send Order message to PFI: $e")
        }
    }

    @Test
    fun `listen for Order Status updates`() {
        val order = TestData.getOrder().apply { sign(customerDid) }
        val orderStatus = TestData.getOrderStatus().apply { sign(pfi) }
        val closeMessage = TestData.getClose().apply { sign(pfi) }

        val orderStatusMsg = orderStatus.data.orderStatus;
        val closeReason = closeMessage.data.reason

        val mockExchange = listOf(orderStatus, closeMessage)
        getExchangeResponse = Json.jsonMapper.writeValueAsString(mapOf("data" to mockExchange))
        server.enqueue(MockResponse().setBody(getExchangeResponse).setResponseCode(HttpURLConnection.HTTP_OK))

        // :snippet-start: listenForOrderStatusKt
        var orderStatusUpdate: String? = ""
        var close: Close? = null

        while (close == null) {
            val exchange = TbdexHttpClient.getExchange(
                pfiDid = order.metadata.to,
                requesterDid = customerDid,
                exchangeId = order.metadata.exchangeId.toString()
            )

            for (message in exchange) {
                when (message) {
                    is OrderStatus -> {
                        // a status update to display to your customer
                        orderStatusUpdate = message.data.orderStatus
                    }
                    is Close -> {
                        // final message of exchange has been written
                        close = message
                    }
                    else -> {}
                }
            }
        }
        // :snippet-end:
        assertEquals(orderStatusMsg, orderStatusUpdate, "Order Status message is incorrect")

        // :snippet-start: getCloseReasonKt
        val isSuccessful = close.data.success
        val reasonForClose = close.data.reason
        // :snippet-end:

        assertTrue(isSuccessful == true, "Close was not successful")
        assertEquals(closeReason, reasonForClose, "Close reason is incorrect")
    }
}
